tests/tests/authentication.ts

```diff
@@ -0,0 +1,128 @@
+ */
+export async function authentication(testState: TestState, implicit = false) {
+  const requirements = await authenticationGet(testState)
+
+  // Construct the signed presentaiton.
+  const seed = Kilt.Utils.Crypto.mnemonicToMiniSecret(mnemonic)
+  const key = deriveEncryptionKeyFromSeed(seed)
+  const authenticationKey = deriveAuthenticationKey(seed)
+
+  const presentation = await Kilt.Credential.createPresentation({
+    credential: CREDENTIAL,
+    signCallback: sign,
+    challenge: requirements.body.content.challenge,
+  })
+
+  async function sign({ data }: Kilt.SignRequestData) {
+    const signature = authenticationKey.sign(data, { withType: false })
+    const keyUri = DID_AUTH_KEY_URL as Kilt.DidResourceUri
+    const keyType = authenticationKey.type
+
+    return {
+      signature,
+      keyUri,
+      keyType,
+    }
+  }
+
+  // Construct the required payload.
+  const payload = {
+    body: { content: [presentation], type: 'credential' },
+    createdAt: 0,
+    sender: '-',
+    receiver: '-',
+    messageId: 'abc123',
+  }
+
+  // Encrypt the payload with the Key Agreement key of the test user's DID Document.
+  const encrypted = Kilt.Utils.Crypto.encryptAsymmetric(
+    JSON.stringify(payload),
+    testState.getOpenDidKeyAgreement().publicKey,
+    key.secretKey,
+  )
+  const postRequestBody: EncryptedMessage = {
+    ciphertext: '0x' + toHex(encrypted.box),
+    senderKeyUri: DID_KEY_AGREEMENT_URL,
+    receiverKeyUri: '-',
+    nonce: '0x' + toHex(encrypted.nonce),
+  }
+
+  // Send the POST Request.
+  const cookie = testState.getCookie()
+  const response = await axios.post(credentialUrl.toString(), postRequestBody, {
+    headers: { Cookie: cookie },
+    validateStatus: () => true,
+  })
+  expect(response.status).toBe(204)
+
+  const url = new URL(response.headers.location)
+  const redirectUri = new URL(REDIRECT_URI)
+  expect(url.origin).toBe(redirectUri.origin)
+  expect(url.pathname).toBe(redirectUri.pathname)
+  if (implicit) {
+    // In the implicit case, the verify the `id_token` which is returned as a fragment identifier.
+    const fragmentIdentifiers = new URLSearchParams(url.hash.replace('#', '?'))
+    expect(fragmentIdentifiers.get('state')).toBe(TestState.STATE)
+    expect(fragmentIdentifiers.get('token_type')).toBe('bearer')
+    expect(fragmentIdentifiers.get('refresh_token')!.length).toBeGreaterThan(10)
+    const idToken = fragmentIdentifiers.get('id_token') as string
+
+    const decodedToken = jsonwebtoken.verify(idToken, JWT_SECRET) as jsonwebtoken.JwtPayload
+    expect(decodedToken.nonce).toBe(TestState.NONCE)
+  } else {
+    // In the Authorization Code Flow case, the code is returned as a query parameter which is stored in the `testState`
+    expect(url.searchParams.get('state')).toBe(TestState.STATE)
+    const code = url.searchParams.get('code')!
+    expect(code.length).toBeGreaterThan(10)
+    testState.setAuthCode(code)
+  }
+}
```

Comment:
It feels we are testing multiple things inside here. It would be nice to separate things here.

Reply:
I extracted functions for better readability, but I'm not sure how to split the tests without duplication a lot of code.

763ba2d3721a